File: bindings/elixir/lib/proven_servers/voip.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

defmodule ProvenServers.Voip do
  @moduledoc """
  VoIP (Voice over IP / SIP) types for the proven-servers ABI.
  
  Mirrors the Idris2 module `VoIPABI.Types` and its type definitions:
  - `Method`       — SIP request methods (13 constructors, tags 0-12)
  - `ResponseCode` — SIP response codes (17 constructors, tags 0-16)
  - `DialogState`  — SIP dialog lifecycle (3 constructors, tags 0-2)
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard SIP port (RFC 3261)."
  @spec sip_port() :: non_neg_integer()
  def sip_port, do: 5060

  @doc "Standard SIP over TLS (SIPS) port (RFC 3261)."
  @spec sips_port() :: non_neg_integer()
  def sips_port, do: 5061

  # ===========================================================================
  # Method (tags 0-12)
  # ===========================================================================

  @typedoc """
  Method types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type method ::
          :invite
          | :ack
          | :bye
          | :cancel
          | :register
          | :options
          | :info
          | :update
          | :subscribe
          | :notify
          | :refer
          | :message
          | :prack

  @method_tags %{
    invite: 0,
    ack: 1,
    bye: 2,
    cancel: 3,
    register: 4,
    options: 5,
    info: 6,
    update: 7,
    subscribe: 8,
    notify: 9,
    refer: 10,
    message: 11,
    prack: 12,
  }

  @tag_to_method Map.new(@method_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Method` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..12, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Voip.method_from_tag(0)
      {:ok, :invite}
  """
  @spec method_from_tag(non_neg_integer()) :: {:ok, method()} | :error
  def method_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 12 do
    {:ok, Map.fetch!(@tag_to_method, tag)}
  end

  def method_from_tag(_tag), do: :error

  @doc """
  Encode a `Method` to the C-ABI tag value.
  """
  @spec method_to_tag(method()) :: non_neg_integer()
  def method_to_tag(val) when is_map_key(@method_tags, val) do
    Map.fetch!(@method_tags, val)
  end

  @doc """
  All `Method` variants in tag order.
  """
  @spec all_methods() :: [method()]
  def all_methods do
    [
      :invite, :ack, :bye, :cancel, :register, :options, :info, :update,
      :subscribe, :notify, :refer, :message, :prack
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  The SIP method name string.
        match self {

  Whether this method creates or modifies a dialog.
  """
  @spec is_dialog_creating?(method()) :: boolean()
  def is_dialog_creating?(val) when val in [:invite, :subscribe], do: true
  def is_dialog_creating?(_val), do: false

  @doc """
  Whether this method is related to session management.
  """
  @spec is_session_related?(method()) :: boolean()
  def is_session_related?(val) when val in [:invite, :ack, :bye, :cancel, :update, :prack], do: true
  def is_session_related?(_val), do: false

  @doc """
  Whether this method is related to event notification.
  """
  @spec is_event_related?(method()) :: boolean()
  def is_event_related?(val) when val in [:subscribe, :notify], do: true
  def is_event_related?(_val), do: false

  # ===========================================================================
  # ResponseCode (tags 0-16)
  # ===========================================================================

  @typedoc """
  ResponseCode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type response_code ::
          :trying
          | :ringing
          | :session_progress
          | :ok
          | :multiple_choices
          | :moved_permanently
          | :moved_temporarily
          | :bad_request
          | :unauthorized
          | :forbidden
          | :not_found
          | :method_not_allowed
          | :request_timeout
          | :busy_here
          | :decline
          | :server_internal_error
          | :service_unavailable

  @response_code_tags %{
    trying: 0,
    ringing: 1,
    session_progress: 2,
    ok: 3,
    multiple_choices: 4,
    moved_permanently: 5,
    moved_temporarily: 6,
    bad_request: 7,
    unauthorized: 8,
    forbidden: 9,
    not_found: 10,
    method_not_allowed: 11,
    request_timeout: 12,
    busy_here: 13,
    decline: 14,
    server_internal_error: 15,
    service_unavailable: 16,
  }

  @tag_to_response_code Map.new(@response_code_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ResponseCode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..16, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Voip.response_code_from_tag(0)
      {:ok, :trying}
  """
  @spec response_code_from_tag(non_neg_integer()) :: {:ok, response_code()} | :error
  def response_code_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 16 do
    {:ok, Map.fetch!(@tag_to_response_code, tag)}
  end

  def response_code_from_tag(_tag), do: :error

  @doc """
  Encode a `ResponseCode` to the C-ABI tag value.
  """
  @spec response_code_to_tag(response_code()) :: non_neg_integer()
  def response_code_to_tag(val) when is_map_key(@response_code_tags, val) do
    Map.fetch!(@response_code_tags, val)
  end

  @doc """
  All `ResponseCode` variants in tag order.
  """
  @spec all_response_codes() :: [response_code()]
  def all_response_codes do
    [
      :trying, :ringing, :session_progress, :ok, :multiple_choices, :moved_permanently,
      :moved_temporarily, :bad_request, :unauthorized, :forbidden, :not_found,
      :method_not_allowed, :request_timeout, :busy_here, :decline, :server_internal_error,
      :service_unavailable
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this is a provisional (1xx) response.
  """
  @spec is_provisional?(response_code()) :: boolean()
  def is_provisional?(val) when val in [:trying, :ringing, :session_progress], do: true
  def is_provisional?(_val), do: false

  @doc """
  Whether this is a success (2xx) response.
  """
  @spec is_success?(response_code()) :: boolean()
  def is_success?(val) when val in [:ok], do: true
  def is_success?(_val), do: false

  @doc """
  Whether this is a redirection (3xx) response.
  """
  @spec is_redirect?(response_code()) :: boolean()
  def is_redirect?(val) when val in [:multiple_choices, :moved_permanently, :moved_temporarily], do: true
  def is_redirect?(_val), do: false

  @doc """
  Whether this is a client error (4xx) response.
  """
  @spec is_client_error?(response_code()) :: boolean()
  def is_client_error?(val) when val in [:bad_request, :unauthorized, :forbidden, :not_found, :method_not_allowed, :request_timeout, :busy_here], do: true
  def is_client_error?(_val), do: false

  @doc """
  Whether this is a server error (5xx) response.
  """
  @spec is_server_error?(response_code()) :: boolean()
  def is_server_error?(val) when val in [:server_internal_error, :service_unavailable], do: true
  def is_server_error?(_val), do: false

  @doc """
  Whether this is a global failure (6xx) response.
  """
  @spec is_global_failure?(response_code()) :: boolean()
  def is_global_failure?(val) when val in [:decline], do: true
  def is_global_failure?(_val), do: false

  # ===========================================================================
  # DialogState (tags 0-2)
  # ===========================================================================

  @typedoc """
  DialogState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type dialog_state :: :early | :confirmed | :terminated

  @dialog_state_tags %{
    early: 0,
    confirmed: 1,
    terminated: 2,
  }

  @tag_to_dialog_state Map.new(@dialog_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `DialogState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..2, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Voip.dialog_state_from_tag(0)
      {:ok, :early}
  """
  @spec dialog_state_from_tag(non_neg_integer()) :: {:ok, dialog_state()} | :error
  def dialog_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 2 do
    {:ok, Map.fetch!(@tag_to_dialog_state, tag)}
  end

  def dialog_state_from_tag(_tag), do: :error

  @doc """
  Encode a `DialogState` to the C-ABI tag value.
  """
  @spec dialog_state_to_tag(dialog_state()) :: non_neg_integer()
  def dialog_state_to_tag(val) when is_map_key(@dialog_state_tags, val) do
    Map.fetch!(@dialog_state_tags, val)
  end

  @doc """
  All `DialogState` variants in tag order.
  """
  @spec all_dialog_states() :: [dialog_state()]
  def all_dialog_states, do: [:early, :confirmed, :terminated]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether media can flow in this state.
  """
  @spec can_carry_media?(dialog_state()) :: boolean()
  def can_carry_media?(val) when val in [:early, :confirmed], do: true
  def can_carry_media?(_val), do: false

  @doc """
  Whether the dialog is active (not terminated).
  """
  @spec is_active?(dialog_state()) :: boolean()
  def is_active?(val) when val in [:terminated], do: false
  def is_active?(_val), do: true

end

File: protocols/proven-voip/ffi/zig/src/voip.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// voip.zig -- Zig FFI implementation of proven-voip.
//
// Implements the SIP/VoIP session state machine with:
//   - 64-slot mutex-protected session pool
//   - Dialog state tracking (Early/Confirmed/Terminated)
//   - SIP method request counting
//   - SIP response code tracking
//   - CSeq management (monotonically increasing)
//   - Registration binding tracking (max 16 per session)
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching VoIPABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching VoIPABI.Types.idr tag assignments)
// =========================================================================

/// SIP methods (ABI tags 0-12).
pub const Method = enum(u8) {
    invite = 0,
    ack = 1,
    bye = 2,
    cancel = 3,
    register = 4,
    options = 5,
    info = 6,
    update = 7,
    subscribe = 8,
    notify = 9,
    refer = 10,
    message = 11,
    prack = 12,
};

/// SIP response codes (ABI tags 0-16).
pub const ResponseCode = enum(u8) {
    trying = 0,
    ringing = 1,
    session_progress = 2,
    ok = 3,
    multiple_choices = 4,
    moved_permanently = 5,
    moved_temporarily = 6,
    bad_request = 7,
    unauthorized = 8,
    forbidden = 9,
    not_found = 10,
    method_not_allowed = 11,
    request_timeout = 12,
    busy_here = 13,
    decline = 14,
    server_internal_error = 15,
    service_unavailable = 16,
};

/// Dialog states (ABI tags 0-2).
pub const DialogState = enum(u8) {
    early = 0,
    confirmed = 1,
    terminated = 2,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent sessions.
const MAX_SESSIONS: usize = 64;

/// Maximum registrations per session.
const MAX_REGISTRATIONS: usize = 16;

/// Maximum Call-ID / contact length.
const MAX_ID_LEN: usize = 256;

/// A registration binding.
const Registration = struct {
    contact: [MAX_ID_LEN]u8,
    contact_len: u32,
    expires: u32,
    active: bool,
};

/// Default (empty) registration.
const empty_registration: Registration = .{
    .contact = [_]u8{0} ** MAX_ID_LEN,
    .contact_len = 0,
    .expires = 0,
    .active = false,
};

/// A VoIP session.
const Session = struct {
    /// Current dialog state.
    state: DialogState,
    /// Call-ID for this session.
    call_id: [MAX_ID_LEN]u8,
    call_id_len: u32,
    /// Current CSeq number (monotonically increasing).
    cseq: u32,
    /// Total requests sent.
    request_count: u32,
    /// Total responses received.
    response_count: u32,
    /// Registration bindings.
    registrations: [MAX_REGISTRATIONS]Registration,
    /// Number of active registrations.
    registration_count: u32,
    /// Whether this session slot is in use.
    active: bool,
};

/// Default (empty) session.
const empty_session: Session = .{
    .state = .early,
    .call_id = [_]u8{0} ** MAX_ID_LEN,
    .call_id_len = 0,
    .cseq = 1,
    .request_count = 0,
    .response_count = 0,
    .registrations = [_]Registration{empty_registration} ** MAX_REGISTRATIONS,
    .registration_count = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

/// Returns the ABI version number. Must match Foreign.abiVersion in Idris2.
pub export fn voip_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new VoIP session. Returns slot index (>=0) or -1 on failure.
/// The session starts in Early state.
pub export fn voip_create(
    call_id_ptr: [*]const u8,
    call_id_len: u32,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (call_id_len == 0 or call_id_len > MAX_ID_LEN) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.call_id[0..call_id_len], call_id_ptr[0..call_id_len]);
            s.call_id_len = call_id_len;
            s.state = .early;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a session, releasing its slot.
pub export fn voip_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

/// Returns the current DialogState tag for a session.
pub export fn voip_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Send a SIP request. Returns 0 on success, 1 on rejection.
pub export fn voip_send_request(slot: c_int, method: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (method > 12) return 1;

    // BYE only allowed from Confirmed state
    if (method == 2 and sessions[idx].state != .confirmed) return 1;
    // Cannot send requests from Terminated state
    if (sessions[idx].state == .terminated) return 1;

    sessions[idx].cseq += 1;
    sessions[idx].request_count += 1;
    return 0;
}

/// Receive a SIP response. Returns 0 on success, 1 on rejection.
pub export fn voip_recv_response(slot: c_int, code: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (code > 16) return 1;
    if (sessions[idx].state == .terminated) return 1;

    sessions[idx].response_count += 1;
    return 0;
}

/// Confirm a dialog. Transitions Early -> Confirmed. Returns 0 on success.
pub export fn voip_confirm(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .early) return 1;

    sessions[idx].state = .confirmed;
    return 0;
}

/// Terminate a dialog. Returns 0 on success, 1 on rejection.
pub export fn voip_terminate(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state == .terminated) return 1;

    sessions[idx].state = .terminated;
    return 0;
}

/// Returns current CSeq number.
pub export fn voip_cseq(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].cseq;
}

/// Register a contact binding. Returns 0 on success, 1 on rejection.
pub export fn voip_register(
    slot: c_int,
    contact_ptr: [*]const u8,
    contact_len: u32,
    expires: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (contact_len == 0 or contact_len > MAX_ID_LEN) return 1;

    for (&sessions[idx].registrations) |*reg| {
        if (!reg.active) {
            @memcpy(reg.contact[0..contact_len], contact_ptr[0..contact_len]);
            reg.contact_len = contact_len;
            reg.expires = expires;
            reg.active = true;
            sessions[idx].registration_count += 1;
            return 0;
        }
    }
    return 1;
}

/// Returns number of active registrations.
pub export fn voip_registration_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].registration_count;
}

/// Returns number of requests sent.
pub export fn voip_request_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].request_count;
}

/// Returns number of responses received.
pub export fn voip_response_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].response_count;
}

/// Check if a dialog state transition is valid.
pub export fn voip_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Early -> Confirmed
    if (from == 0 and to == 2) return 1; // Early -> Terminated
    if (from == 1 and to == 2) return 1; // Confirmed -> Terminated
    return 0;
}
